### PR TITLE
Booking event link

### LINKED ISSUE #3694: Assigning booking to event
Whenever we're confirming a booking we should show a dropdown with overlapping events owned by the user.

### PR DESCRIPTION
**Pull request by meluru**
Closes  #3694

### CODE REVIEW

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 1]
```diff
@@ -290,6 +299,52 @@ class BookRoomModal extends React.Component {
         );
     }
 
+    renderEventItem = (event) => {
+        const start = `${moment(event.startDt).format('L')} ${moment(event.startDt).format('LT')}`;
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: Won't this do the same?

```js
const start = moment(event.startDt).format('L LT');
```

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 2]
```diff
@@ -290,6 +299,52 @@ class BookRoomModal extends React.Component {
...
+        }
+
+        return (
+            <Segment>
+                <h3><Icon name="chain" />{Translate.string('Event')}</h3>
+                <div styleName="events-segment-description">
+                    <PluralTranslate count={relatedEvents.length}>
+                        <Singular>
+                            You have an event taking place in this room overlapping with booking times.
+                            If you are booking this room for this event, please select it below.
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: I think "...the room for this event..." sounds better than having "this" there twice

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 3]
```diff
@@ -404,14 +462,17 @@ export default connect(
         favoriteUsers: userSelectors.getFavoriteUsers(state),
         userFullName: userSelectors.getUserFullName(state),
         availability: state.bookRoom.bookingForm.availability,
+        relatedEvents: state.bookRoom.bookingForm.relatedEvents,
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: both this and availability should use selectors.. let's add `getBookingFormAvailability` and `getBookingFormRelatedEvents`?

File: indico/modules/rb_new/operations/rooms.py [hunk 4]
```diff
@@ -198,3 +201,18 @@ def search_for_rooms(filters, availability=None):
...
+def get_room_events(room, start_dt, end_dt, repeat_frequency, repeat_interval):
+    occurrences = ReservationOccurrence.create_series(start_dt, end_dt, (repeat_frequency, repeat_interval))
+    excluded_categories = rb_settings.get('excluded_categories')
+    events = (Event.query
+              .filter(~Event.is_deleted,
+                      Event.own_room == room,
+                      db.or_(Event.happens_between(as_utc(occ.start_dt), as_utc(occ.end_dt)) for occ in occurrences),
+                      Event.timezone == config.DEFAULT_TIMEZONE,
+                      db.and_(Event.category_id != cat['id'] for cat in excluded_categories))
+              .options(subqueryload('acl_entries'))).all()
```
Comment (ThiefMaster) [inline indico/modules/rb_new/operations/rooms.py]:
```py
.options(subqueryload('acl_entries'))
.all())
```

File: indico/modules/rb_new/operations/rooms.py [hunk 5]
```diff
@@ -198,3 +201,18 @@ def search_for_rooms(filters, availability=None):
...
+    user_events = [event for event in events if [entry for entry in event.acl_entries
```
Comment (ThiefMaster) [inline indico/modules/rb_new/operations/rooms.py]: Please don't iterate over `acl_entries` manually - we have shortcuts for this, i.e. `event.can_manage(session.user)`

If this is for performance reasons to skip group-based access, then do the filtering directly in the query (and remove the subqueryload:

```py
                      Event.acl_entries.any(db.and_(EventPrincipal.type == PrincipalType.user,
                                                    EventPrincipal.user_id == session.user.id,
                                                    EventPrincipal.full_access))
```

File: indico/modules/rb_new/schemas.py [hunk 6]
```diff
@@ -83,6 +83,13 @@ class ReservationEventDataSchema(Schema):
     can_access = Function(lambda event: event.can_access(session.user))
 
 
+class ReservationUserEventsSchema(Schema):
```
Comment (ThiefMaster) [inline indico/modules/rb_new/schemas.py]: `ReservationUserEventSchema`

File: indico/modules/rb_new/operations/rooms.py (hunk 5 above)
Comment (meluru) [reply indico/modules/rb_new/operations/rooms.py]: We discussed on irc to skip group-based access and I couldn't figure out the proper query :D' thanks

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 7]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
         );
     }
 
+    renderEventItem = (event) => {
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: `getEventOption`? A `render*` method usually returns JSX

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 8]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
...
+        const start = `${moment(event.startDt).format('L LT')}`;
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: no need for the template string wrapper

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 9]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
...
+            return;
+        }
+
+        return (
+            <Segment>
+                <h3><Icon name="chain" />{Translate.string('Event')}</h3>
+                <div styleName="events-segment-description">
+                    <PluralTranslate count={relatedEvents.length}>
+                        <Singular>
+                            You have an event taking place in this room overlapping with booking times.
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: "overlapping with booking times" sounds very weird... maybe just "You have an event taking place in this room" without actually mentioning the overlap? we show the event times anyway, right?

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 10]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
...
+                            If you are booking the room for this event, please select it below.
+                        </Singular>
+                        <Plural>
+                            You have events taking place in this room overlapping with booking times.
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: same

File: indico/modules/rb_new/operations/rooms.py [hunk 11]
```diff
@@ -198,3 +203,19 @@ def search_for_rooms(filters, availability=None):
                                             db.or_(booking_limit_days.is_(None),
                                                    selected_period_days <= booking_limit_days))))
     return query
+
+
+def get_room_events(room, start_dt, end_dt, repeat_frequency, repeat_interval):
+    occurrences = ReservationOccurrence.create_series(start_dt, end_dt, (repeat_frequency, repeat_interval))
+    excluded_categories = rb_settings.get('excluded_categories')
+    events = (Event.query
```
Comment (ThiefMaster) [inline indico/modules/rb_new/operations/rooms.py]: return (...)

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx (hunk 9 above)
Comment (meluru) [reply indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: yeah, i guess it will be enough

**Comment by meluru** [comment]
Updated

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 12]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
         );
     }
 
+    getEventOption = (event) => {
+        const start = moment(event.startDt).format('L LT');
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: You could use `serializeDate` here ;)

File: indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx [hunk 13]
```diff
@@ -290,6 +300,52 @@ class BookRoomModal extends React.Component {
...
+        const sameDate = moment(event.startDt).isSame(event.endDt, 'date');
+        const endTime = moment(event.endDt).format('LT');
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: The same as above.

Comment (meluru) [reply indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: shouldn't it be `serializeTime` here? I know it does the same as `serializeDate` when I pass a format parameter, but the naming could make it confusing to read.

Comment (mic4ael) [reply indico/modules/rb_new/client/js/modules/bookRoom/BookRoomModal.jsx]: Yeah, works fine as well. Even better ;)

**Pull request merged**